### PR TITLE
Support for complex ORDER BY expressions referencing projection output

### PR DESCRIPTION
**Pull request by sopel39**
This makes various TPCDS queries to execute unmodified

FYI: @martint

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -978,6 +978,13 @@ public void testOrderByExpressionWithLimit()
     }
 
     @Test
+    public void testOrderByWithProjectionColumnReference()
+    {
+        assertQueryOrdered("SELECT a*2 AS b FROM (VALUES -1, 0, 2) t(a) ORDER BY b*-1", "VALUES (4), (0), (-2)");
+        assertQueryOrdered("SELECT a*2 AS b FROM (VALUES -1, 0, 2) t(a) ORDER BY b", "VALUES (-2), (0), (4)");
```
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: I think you need to add more testing:
```
SELECT a* -1 AS a FROM (VALUES -1, 0, 2) t(a) ORDER BY a*-1
SELECT a*-1 AS a FROM (VALUES -1, 0, 2) t(a) ORDER BY t.a*-1
SELECT a*-1 FROM (VALUES -1, 0, 2) t(a) ORDER BY a*-1
SELECT a*-1 FROM (VALUES -1, 0, 2) t(a) ORDER BY t.a*-1
```

Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: More tests ;)
```
SELECT a, a* -1 AS a FROM (VALUES -1, 0, 2) t(a) ORDER BY a; -- expected failure
SELECT a, a* -1 AS a FROM (VALUES -1, 0, 2) t(a) ORDER BY t.a;
SELECT a, a* -2 AS b FROM (VALUES -1, 0, 2) t(a) ORDER BY a + b
SELECT a as b, a* -2 AS a FROM (VALUES -1, 0, 2) t(a) ORDER BY a + b;
SELECT a* -2 AS a FROM (VALUES -1, 0, 2) t(a) ORDER BY a + t.a;
SELECT (SELECT 1 as a) FROM (VALUES -1, 0, 2) t(a) ORDER BY a; -- expected failure?
```

Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: AFAIK you can write "VALUES -1, 0, 2" for h2 check query (no need for '(', ')')

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -1168,10 +1156,8 @@ else if (expression instanceof LongLiteral) {
 
                     orderByExpression = outputExpressions.get(field);
                 }
-
-                // otherwise, just use the expression as is
-                if (orderByExpression == null) {
-                    orderByExpression = expression;
+                else {
+                    orderByExpression = ExpressionTreeRewriter.rewriteWith(new OrderByExpressionRewriter(byAlias), expression);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: IMO instead of using `byAlias` you should use outputScope

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: It may blow up, so please just give it a try

**Comment by kokosing** [comment]
Add to commit message that this commit fixes: https://github.com/prestodb/presto/issues/4698

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Scope doesn't provide information about expressions

**Comment by sopel39** [comment]
applied comments

**Comment by martint** [comment]
Instead of referencing the issue in the commit message, describe the actual problem.

**Comment by sopel39** [comment]
rebased & amended commits so it is more descriptive.

**Comment by martint** [comment]
I made a couple of minor changes and merged it. Thanks!

**Pull request closed without merging**